---
title: "Subseting, Filtering, LD_Prunnig and PCA"
author: "Jaskaran S. Gill"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    fig_height: 9
    fig_width: 16
    highlight: haddock
    theme: cosmo
    toc: yes
    toc_float: yes
  word_document:
    toc: yes
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE , message = FALSE)
```

## Code Subsetting and Filtering 
```{bash, eval=FALSE}
#!/bin/bash
#SBATCH --get-user-env
#SBATCH --clusters=biohpc_gen
#SBATCH --partition=biohpc_gen_normal
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=4763mb
#SBATCH --time=1-12:00:00
#SBATCH -J Subsetting_&_Filtering
#SBATCH --error=%J.err


####The VCF file as an Input, and Name of the sample to keep in the text format in the "samples_to_keep"####################################################

input=$1
sample_to_keep=$2
VCF="/dss/dsshome1/09/ra78pec/data/subset_sample.vcf.gz" #Location of the output VCF File

######Ativating the environment containing the vcftools####################################################################################################

source /dss/dsshome1/09/ra78pec/.bashrc #Bashrc Location in my server
source activate gill #Activating the environment which has the vcftools

###########################################################################################################################################################

vcftools --gzvcf $input \
         -keep $sample_to_keep --recode --stdout | gzip -c > /dss/dsshome1/09/ra78pec/data/subset_sample.vcf.gz

############################################################################################################################################################

####Setting the criteria for filteirng #####################################################################################################################

MAF=0.02 
MISS=0.8
QUAL=20
MIN_DEPTH=5
MAX_DEPTH=35
VCF_OUT="/dss/dsshome1/09/ra78pec/data/subset_filtered.vcf.gz" ##name and location of the output filtered VCF



####### perform the filtering with vcftools#################################################################################################################

vcftools --gzvcf $VCF \
         --remove-indels --maf $MAF --max-missing $MISS --minQ $QUAL \
         --min-meanDP $MIN_DEPTH --max-meanDP $MAX_DEPTH \
         --minDP $MIN_DEPTH --maxDP $MAX_DEPTH --recode --stdout | gzip -c > /dss/dsshome1/09/ra78pec/data/subset_filtered.vcf.gz




################################################################################################################################################################### Accessing the Quality of the Data ##################################################################################################################


OUT="/dss/dsshome1/09/ra78pec/results/quality_after_filtering/subset_quality" ##Folder and prefix of the Output files

bcftools index $VCF_OUT

vcftools --gzvcf $VCF_OUT --freq2 --out $OUT.2 --max-alleles 2
vcftools --gzvcf $VCF_OUT --depth --out $OUT
vcftools --gzvcf $VCF_OUT --site-mean-depth --out $OUT
vcftools --gzvcf $VCF_OUT --site-quality --out $OUT
vcftools --gzvcf $VCF_OUT --missing-indv --out $OUT
vcftools --gzvcf $VCF_OUT --missing-site --out $OUT
vcftools --gzvcf $VCF_OUT --het --out $OUT

############################################################################################################################################################

##Plotting the quality graphs###############################################################################################################################
###Deacivaitng the gill environment as the base had R which is required for this analysis###################################################################


outdir="/dss/dsshome1/09/ra78pec/results/quality_after_filtering/results" ##directory where we want the pdf of the Quality graphs
conda deactivate
cd $outdir

Rscript /dss/dsshome1/09/ra78pec/run/plot.R

cd /dss/dsshome1/09/ra78pec


########################LD_Prunning#########################################################################################################################

source /dss/dsshome1/09/ra78pec/.bashrc
source activate gill
vcf_prunned="/dss/dsshome1/09/ra78pec/data/subset_filtered_prunned.vcf"
#######Finding the sites to Keep
plink --vcf $VCF_OUT \
      --const-fid \
      --allow-extra-chr \
      --allow-no-sex \
      --keep-allele-order \
      --chr-set 37 no-xy \
      --set-missing-var-ids @:# \
      --indep-pairwise 50 10 0.1 \
      --out /dss/dsshome1/09/ra78pec/results/LD_Prunning/subset-filter

#######Filtering the sites to keep from the VCF File
plink --vcf $VCF_OUT \
      --keep-allele-order \
      --allow-extra-chr \
      --allow-no-sex  \
      --const-fid \
      --set-missing-var-ids @:# \
      --chr-set 37 no-xy \
      --extract /dss/dsshome1/09/ra78pec/results/LD_Prunning/subset-filter.prune.in \
      --recode vcf \
      --out $vcf_prunned


############################################################################################################################################################
############################################################################################################################################################
############Getting Eigenvalues for the PCA

plink --vcf $vcf_prunned \
      --keep-allele-order \
      --allow-extra-chr \
      --allow-no-sex  \
      --const-fid \
      --set-missing-var-ids @:# \
      --chr-set 37 no-xy \
      --pca \
      --out /dss/dsshome1/09/ra78pec/results/PCA_after_LD/pca ##Location whereyou want to save your file


#############format changing to gds#########################################################################################################################
conda deactivate

Rscript --vanilla /dss/dsshome1/09/ra78pec/run/gds.R

##gds format is required for the Certain R libraries
#########################Plink format file for the Admixture################################################################################################

source /dss/dsshome1/09/ra78pec/.bashrc
source activate gill

plink --vcf $vcf_prunned \
      --const-fid \
      --allow-extra-chr \
      --allow-no-sex \
      --autosome \
      --keep-allele-order \
      --chr-set 37 no-xy \
      --set-missing-var-ids @:# \
      --recode --make-bed \
      --out /dss/dsshome1/09/ra78pec/data/subset


```
## Plotting the Quality

```{r}
library(ggplot2)
library(tidyr)
library(dplyr)
library(readr)


## -----------------------------------------------------------------------------------------------------------------------------------------
#change before running
name <- "/home/gill/Project_2022/PCA_analysis/quality_after_filtering/subset_quality"


## -----------------------------------------------------------------------------------------------------------------------------------------
var_qual <- read_delim(paste0(name,".lqual"), delim = "\t", col_names = c("chr", "pos", "qual"), skip = 1)

a <- ggplot(var_qual, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))

summary(var_qual$qual)



## -----------------------------------------------------------------------------------------------------------------------------------------
var_depth <- read_delim(paste0(name,".ldepth.mean"), delim = "\t", col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)
a <- ggplot(var_depth, aes(mean_depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))

a + theme_light() + xlim(0, 35)

summary(var_depth$mean_depth)



## -----------------------------------------------------------------------------------------------------------------------------------------
var_miss <- read_delim(paste0(name,".lmiss"), delim = "\t", col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(var_miss, aes(fmiss)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))
summary(var_miss$fmiss)


## -----------------------------------------------------------------------------------------------------------------------------------------
var_freq <- read_delim(paste0(name,".2.frq"), delim = "\t", col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1)
var_freq$maf <- var_freq %>% select(a1, a2) %>% apply(1, function(z) min(z))
a <- ggplot(var_freq, aes(maf)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))
summary(var_freq$maf)



## -----------------------------------------------------------------------------------------------------------------------------------------
summary(var_freq$maf)


## -----------------------------------------------------------------------------------------------------------------------------------------
ind_depth <- read_delim(paste0(name,".idepth"), delim = "\t", col_names = c("ind", "nsites", "depth"), skip = 1)
a <- ggplot(ind_depth, aes(depth)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))


## -----------------------------------------------------------------------------------------------------------------------------------------
summary(ind_depth$depth)
ind_depth %>% filter_at(vars(3:3), any_vars(. < 5))



## -----------------------------------------------------------------------------------------------------------------------------------------
ind_depth %>% filter_at(vars(3:3), any_vars(. < 5))


## -----------------------------------------------------------------------------------------------------------------------------------------
ind_miss  <- read_delim(paste0(name,".imiss"), delim = "\t", col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(ind_miss, aes(fmiss)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))


## -----------------------------------------------------------------------------------------------------------------------------------------
ind_miss %>% filter_at(vars(5:5), any_vars(. > 0.5))


## -----------------------------------------------------------------------------------------------------------------------------------------
ind_het  <- read_delim(paste0(name,".het"), delim = "\t", col_names = c("ind", "o.hom", "e.hom", "nsites", "f"), skip = 1)
ind_het <- ind_het %>% 
  rowwise() %>%
  add_column(o.het =ind_het$nsites-ind_het$o.hom)

ind_het <- ind_het %>% 
  rowwise() %>%
  add_column(o.het.pro =ind_het$o.het/ind_het$nsites)

a <- ggplot(ind_het, aes(o.het.pro)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))

ind_het %>% filter_at(vars(7:7), any_vars(. < 0.16))

a <- ggplot(ind_het, aes(f)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() + theme(axis.text=element_text(size=12),
                          axis.title=element_text(size=14,face="bold"))

ind_het %>% filter_at(vars(5:5), any_vars(. > 0.6))



## -----------------------------------------------------------------------------------------------------------------------------------------
###checking if missingness influences heterozygosity
join <- left_join(ind_miss, ind_het, "ind")

a <- ggscatter(join, x = 'fmiss', y = 'o.het.pro', add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson") +
  geom_point(shape = "circle", size = 1.5, colour = "#112446") +
  theme_minimal()

a
a + theme_light() + xlim(0, 0.1)
```

