---
title: "VCF data statistics"
author: "Sarah Mueller"
date: "2022-12-13"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    theme: lumen
    code_folding: hide
---

Used file: all.vcf.gz

We ran the next script to obtain the stats:

```{bash, qualitycheck_script ,eval=FALSE}

#!/bin/bash
#SBATCH --get-user-env
#SBATCH --clusters=biohpc_gen
#SBATCH --partition=biohpc_gen_normal
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=4763mb
#SBATCH --time=12:00:00
#SBATCH -J vcf.check
#SBATCH --error=/dss/dsshome1/04/ra96her/shallow_project/Logs/check_vcftools.err
#SBATCH --out=/dss/dsshome1/04/ra96her/shallow_project/Logs/check_vcftools.out


##### Shallow WGS - RRBS Project
#Perform a quality check of the vcf 
#Written by: Sarah Muller, modified by Paulina Nunez-Valencia <paulina.nunez.ilp@gmail.com>
#Modification date: 2022-12-09
### Use: sbatch 0_checkvcf_vcftools.sh  <vcf> <workdir> <prefix> 
###NOTE: if you want to modify filters, please do it manually

#___________ PREP _______________

vcf=$1
outdir=$2/$3
prefix=$3
out=$outdir/$prefix

#___________ MAIN _______________

mkdir -p $outdir

# perform quality check
bcftools index $vcf
bcftools view -H $vcf | wc -l

vcftools --gzvcf $vcf --freq2 --out $out --max-alleles 2
vcftools --gzvcf $vcf --freq2 --out $out
vcftools --gzvcf $vcf --depth --out $out
vcftools --gzvcf $vcf --site-mean-depth --out $out
vcftools --gzvcf $vcf --site-quality --out $out
vcftools --gzvcf $vcf --missing-indv --out $out
vcftools --gzvcf $vcf --missing-site --out $out
vcftools --gzvcf $vcf --het --out $out

```


```{r, setup, echo=F, message=FALSE, warning=FALSE}
# load libraries --------

listOfPackages <- c("ggplot2","tidyverse","dplyr", "plotly")
for (i in listOfPackages){
  if(! i %in% installed.packages()){
    install.packages(i, dependencies = TRUE)
  }
  require(i,character.only=TRUE)
}

# def variables --------
wd <- "/dss/dsslegfs01/pr53da/pr53da-dss-0034/projects/2022_SwallowWGS/Results/1_Pre-processing_qualitycheck/rustica.allsites"
prefix <- "rustica.allsites"

setwd(wd)

# def functions --------

```


## Key info

|       |  |
| ----------- | ----------- |
| File size      | 12 G       |
| Indivduals present   | 27 / 27  |
| Number of variants   | 63,662,214   |


## Variant based statistics

### Variant quality

Measure of how much confidence we have in our variant calls. 

```{r, warning=FALSE, message=FALSE}
var_qual <- read_delim(paste0(prefix,".lqual"), delim = "\t",
           col_names = c("chr", "pos", "Quality"), skip = 1)

var_qual.plot <- ggplot(var_qual, aes(Quality)) + 
                  geom_density(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw()

ggplotly(var_qual.plot)
```

We can make sure that our data contains high confidence calls.

### Variant mean depth

Number of reads that have mapped to this position. 

```{r, warning=FALSE, message=FALSE}
var_depth <- read_delim(paste0(prefix,".ldepth.mean"), delim = "\t",
           col_names = c("chr", "pos", "Mean_depth", "Var_depth"), skip = 1)

var_depth.plot <- ggplot(var_depth, aes(Mean_depth)) + 
                  geom_density(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw()

ggplotly(var_depth.plot)

summary(var_depth$Mean_depth)
```

We can clearly observe that our data has a mean depth of 16x. Where most variants have a depth of 10-23x, with few outliers.

### Variant missingness

Measure of how many individuals lack a genotype at a call site. 


```{r, warning=FALSE, message=FALSE}

var_miss <- read_delim(paste0(prefix,".lmiss"), delim = "\t",
                       col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)

var_miss.plot <- ggplot(var_miss, aes(fmiss)) + 
                  geom_density(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw()

ggplotly(var_miss.plot)

summary(var_miss$fmiss)
```

We can observe that indeed, we only have max 10% missing data. Indicating that most individuals have a call at almost every site. We have a mean of 6% of missingness in our data set. 

### Minor allele frequency

Distribution of allele frequencies.

```{r, warning=FALSE, message=FALSE}
var_depth <- read_delim(paste0(prefix,".ldepth.mean"), delim = "\t",
           col_names = c("chr", "pos", "Mean_depth", "Var_depth"), skip = 1)


var_freq <- read_delim(paste0(prefix,".frq"), delim = "\t",
                       col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1)

# find minor allele frequency
var_freq$maf <- var_freq %>% select(a1, a2) %>% apply(1, function(z) min(z))

var_freq.plot <- ggplot(var_freq, aes(maf)) + 
                  geom_density(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw()

ggplotly(var_freq.plot)

summary(var_freq$maf)
```

## Individual based statistics

### Mean depth per individual

Distribution of mean depth among individuals.

```{r, warning=FALSE, message=FALSE}

ind_depth <- read_delim(paste0(prefix,".idepth"), delim = "\t",
                        col_names = c("ind", "nsites", "depth"), skip = 1)
ind_depth.plot <- ggplot(ind_depth, aes(depth)) + 
                  geom_histogram(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw()

ggplotly(ind_depth.plot)

#Per individual
ind_depth2 <- ind_depth                                                
ind_depth2$ind<- factor(ind_depth2$ind,                                    
                  levels = ind_depth2$ind[order(ind_depth2$depth, decreasing = FALSE)])

ind_depth.plot.ind <- ggplot(ind_depth2, aes(x= ind, y= depth))  + 
                  geom_bar(stat="identity", fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw() + 
                  coord_flip()

ggplotly(ind_depth.plot.ind)

```

### Proportion of missing data per individual
```{r, warning=FALSE, message=FALSE}

ind_miss  <- read_delim(paste0(prefix,".imiss"), delim = "\t",
                        col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), skip = 1)
#Overall picture
ind_miss.plot.hist <- ggplot(ind_miss, aes(fmiss)) + 
                         geom_histogram(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                        theme_bw()
ggplotly(ind_miss.plot.hist)

#Per individual
ind_miss2 <- ind_miss                                                
ind_miss2$ind<- factor(ind_miss2$ind,                                    
                  levels = ind_miss2$ind[order(ind_miss2$fmiss, decreasing = FALSE)])

ind_miss.plot <- ggplot(ind_miss2, aes(x= ind, y= fmiss))  + 
                  geom_bar(stat="identity", fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw() + 
                  coord_flip()

ggplotly(ind_miss.plot)

```

### Heterozygosity and inbreeding coefficient per individual

```{r, warning=FALSE, message=FALSE}
ind_het <- read_delim(paste0(prefix,".het"), delim = "\t",
           col_names = c("ind","ho", "he", "nsites", "f"), skip = 1)

ind_het.plot <- ggplot(ind_het, aes(f)) + 
                  geom_histogram(fill = "#DA4E2F", colour = "black", alpha = 0.4) +
                  theme_bw()

ggplotly(ind_het.plot)
```


> NOTE: This code was taken from [here](https://speciationgenomics.github.io/filtering_vcfs/)